Api/Services/ThreadService.cs

```diff
@@ -206,4 +207,129 @@ public async Task<Result<ThreadVM>> GetThreadAsync(int threadId, string userId)
+            AuthorId = userId,
+            MessageThreadId = threadId
+        };
+
+        var user = await dbContext.Users
+        .Where(u => u.Id == userId)
+        .FirstOrDefaultAsync();
+
+        messageThread.Messages.Add(message);
+        await dbContext.SaveChangesAsync();
+
+        messageThread.Messages.Add(message);
+
+        return new MessageVM
+        {
+            Id = message.Id,
+            Contents = message.Contents,
+            DateSent = message.DateSent,
+            DateRead = message.DateRead,
+            AuthorsFirstName = user.FirstName,
+            AuthorsLastName = user.LastName,
+            MessageThreadId = message.MessageThreadId
+        };
+    }
+
+    /// <summary>
+    /// Get threads the logged-in user participates in
+    /// </summary>
+    /// <param name="threadId">id of thread</param>
+    /// <param name="userId">id of thread</param>
+    /// <param name="messageId">id of a message to dispaly</param>
+    /// <param name="perPage">Records per page</param>
+    /// <returns>returns paginated messages starting with provided message id </returns>
+   
+    public async Task<Result<Pagination<MessageVM>>> GetThreadMessagesByIdAsync(int threadId, String userId, int messageId, int perPage)
+    {
+        var messageThread = await dbContext.MessageThreads
+            .Include(t => t.Participants)
+            .FirstOrDefaultAsync(t => t.Id == threadId);
+
+        if (messageThread == null)
+        {
+            return new ValidationFailure
+            {
+                PropertyName = null,
+                ErrorMessage = "Thread not found or you are not a participant.",
+                ErrorCode = ErrorCodes.NotFound
+            };
+        }
+
+        if (!messageThread.Participants.Any(p => p.Id == userId))
+        {
+            return new ValidationFailure
+            {
+                PropertyName = null,
+                ErrorMessage = "Thread not accasable to provided user.",
+                ErrorCode = ErrorCodes.AccessDenied
+            };
+        }
+
+        var query = dbContext.Messages
+            .Include(m => m.Author)
+            .Where(m => m.MessageThreadId == threadId && m.Id > messageId)
+            .OrderByDescending(m => m.DateSent)
+            .Select(m => new MessageVM
+            {
+                Id = m.Id,
+                Contents = m.Contents,
+                DateSent = m.DateSent,
+                DateRead = m.DateRead,
+                AuthorsFirstName = m.Author.FirstName,
+                AuthorsLastName = m.Author.LastName,
+                MessageThreadId = m.MessageThreadId
+            });
+
+        if (query.Count() == 0)
+        {
+            return new ValidationFailure
+            {
+                PropertyName = null,
+                ErrorMessage = "No messages of provided thread found",
+                ErrorCode = ErrorCodes.NotFound
+            };
+        }
+
+        return await query.PaginateAsync(0, perPage, []);
```

Comment:
To PaginateAsync zrobi to co chcemy, ale to jest overkill, można po prostu dodać `.Take(perPage)` na koniec query i zwrócić zwykłą listę